### PR TITLE
Clean mutable tables

### PR DESCRIPTION
**Pull request by kokosing**
Clean mutable tables

Changing the way how mutables are cleaned. Now it works as follows:

 - on tempto start all found mutable tables are removed
 - on each test success its mutable tables are removed

Note that when test fails mutable tables are note removed, thanks to
that it easier to reproduce test failure manually in order to fix it.

Test Plan: manual

Reviewers: losipiuk, cwallin

### CODE REVIEW

**Comment by losipiuk** [comment]
s/note/not/

**Comment by losipiuk** [comment]
s/, thanks to/. Thanks to/

File: tempto-core/src/main/java/com/teradata/tempto/fulfillment/TestStatus.java [hunk 1]
```diff
@@ -12,12 +12,9 @@
...
  */
 
-package com.teradata.tempto.internal.uuid;
+package com.teradata.tempto.fulfillment;
 
-/**
- * Generates random String UUIDs.
- */
-public interface UUIDGenerator
+public interface TestStatus
```
Comment (losipiuk) [inline tempto-core/src/main/java/com/teradata/tempto/fulfillment/TestStatus.java]: why not just enum?

File: tempto-core/src/main/java/com/teradata/tempto/fulfillment/table/TableManager.java [hunk 2]
```diff
@@ -43,13 +44,16 @@
...
-    default TableInstance<T> createMutable(T tableDefinition, State state) {
+    default TableInstance<T> createMutable(T tableDefinition, State state)
+    {
         return createMutable(tableDefinition, state, tableDefinition.getName());
     }
 
     TableInstance<T> createMutable(T tableDefinition, State state, String name);
 
-    void dropAllTables();
+    void dropMutableTables();
```
Comment (losipiuk) [inline tempto-core/src/main/java/com/teradata/tempto/fulfillment/table/TableManager.java]: what is difference between these two methods. Can't guess based on name.

File: tempto-core/src/main/java/com/teradata/tempto/fulfillment/table/TablesState.java [hunk 3]
```diff
@@ -86,6 +86,11 @@ public DatabaseTableInstanceMap getDatabaseTableInstanceMap(String database)
         return databaseTableInstances.get(database);
     }
 
+    public boolean hasDatabaseTableInstanceMap(String database)
```
Comment (losipiuk) [inline tempto-core/src/main/java/com/teradata/tempto/fulfillment/table/TablesState.java]: Exposing internals is weird.

File: tempto-core/src/main/java/com/teradata/tempto/internal/fulfillment/table/AbstractTableManager.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    {
+        this.queryExecutor = requireNonNull(queryExecutor, "queryExecutor is null");
+        this.tableNameGenerator = requireNonNull(tableNameGenerator, "tableNameGenerator is null");
+    }
+
+    @Override
+    public void dropAllMutableTables()
+    {
+        try {
+            DatabaseMetaData metaData = queryExecutor.getConnection().getMetaData();
```
Comment (losipiuk) [inline tempto-core/src/main/java/com/teradata/tempto/internal/fulfillment/table/AbstractTableManager.java]: So we depend on fact that query executor is JDBC based. Maybe call this class AbstractJdbcTableManager

File: tempto-core/src/main/java/com/teradata/tempto/internal/fulfillment/table/TableNameGenerator.java [hunk 5]
```diff
@@ -14,25 +14,25 @@
...
-
-import javax.inject.Inject;
-
-import static com.google.common.base.Preconditions.checkNotNull;
+import java.util.UUID;
 
 public class TableNameGenerator
 {
-    private final UUIDGenerator uuidGenerator;
+    private static final String MUTABLE_TABLE_NAME_PREFIX = "mut_";
```
Comment (losipiuk) [inline tempto-core/src/main/java/com/teradata/tempto/internal/fulfillment/table/TableNameGenerator.java]: make it "tempto_mut"

File: tempto-core/src/main/java/com/teradata/tempto/internal/initialization/TestInitializationListener.java [hunk 6]
```diff
@@ -87,6 +88,9 @@
             MutableTablesFulfiller.class
     );
 
+    private static final TestStatus FAILURE = () -> true;
```
Comment (losipiuk) [inline tempto-core/src/main/java/com/teradata/tempto/internal/initialization/TestInitializationListener.java]: define those as public constants in TestStatus

**Comment by losipiuk** [comment]
looks good modulo comments

File: tempto-core/src/main/java/com/teradata/tempto/internal/fulfillment/table/AbstractTableManager.java (hunk 4 above)
Comment (kokosing) [inline tempto-core/src/main/java/com/teradata/tempto/internal/fulfillment/table/AbstractTableManager.java]: no.... all queryExecutors return connection so it is general. It works for hive also.

**Comment by cawallin** [comment]
looks good, thanks!

File: tempto-core/src/main/java/com/teradata/tempto/internal/fulfillment/table/MutableTablesFulfiller.java [hunk 7]
```diff
@@ -45,4 +52,19 @@ protected TableInstance createTable(TableManager tableManager, MutableTableRequi
     {
         return tableManager.createMutable(tableRequirement.getTableDefinition(), tableRequirement.getState(), tableRequirement.getName());
     }
+
+    @Override
+    public final void cleanup(TestStatus testStatus)
+    {
+        checkState(mutableTablesState != null, "No mutable tables, call fullfil method first");
```
Comment (akshatnair) [inline tempto-core/src/main/java/com/teradata/tempto/internal/fulfillment/table/MutableTablesFulfiller.java]: s/fullfil/fulfill

**Pull request merged**